src/viper/trans.ml

```diff
@@ -323,40 +324,39 @@ and stmt ctxt (s : M.exp) : seqn =
        | [] ->
          fun _ x -> x
        | ConcurrencyS ("1", _, cond) :: _ ->
-         let (!!) p = !!! (cond.at) p in
+         let (!?) p = !!! (cond.at) p in
          let zero, one = intLitE Source.no_region 0, intLitE Source.no_region 1 in
          fun ctxt x ->
-           let ghost_fld () = !!(FldAcc (self ctxt cond.at, !!id)) in
-           let between = !!(AndE (!!(LeCmpE (zero, ghost_fld ())), !!(LeCmpE (ghost_fld (), one)))) in
-           let is_one = !!(EqCmpE (ghost_fld (), one)) in
-           !!(AndE (x, !!(AndE (between, !!(Implies (is_one, cond.it (exp ctxt)))))))
+           let ghost_fld () = !?(FldAcc (self ctxt cond.at, !?id)) in
+           let between = !?(AndE (!?(LeCmpE (zero, ghost_fld ())), !?(LeCmpE (ghost_fld (), one)))) in
+           let is_one = !?(EqCmpE (ghost_fld (), one)) in
+           !?(AndE (x, !?(AndE (between, !?(Implies (is_one, cond.it (exp ctxt)))))))
        | _ -> unsupported e.at (Arrange.exp e) in
      ctxt.ghost_conc := mk_c :: !(ctxt.ghost_conc);
-     !!  ([],
-          [ !!(FieldAssignS(
-              (self ctxt Source.no_region, !!id),
-              (!!(AddE(!!(FldAcc (self ctxt (s.at), !!id)),
-                       intLitE Source.no_region 1)))));
-            !!(ExhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                !!(MacroCall("$Inv", self ctxt at))))));
-            !!(SeqnS (
-                let (!!) p = !!! at p in
-                !!([],
-                   [
-                     !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                    !!(AndE(!!(MacroCall("$Inv", self ctxt at)),
-                                            !!(GtCmpE(!!(FldAcc (self ctxt at, !!id)),
-                                                 intLitE Source.no_region 0))))))));
-                     !!(FieldAssignS(
-                            (self ctxt at, !!id),
-                            (!!(SubE(!!(FldAcc (self ctxt at, !!id)),
-                                     intLitE at 1)))));
-                     !!! (e.at) (SeqnS stmts);
-                     !!(ExhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                         !!(MacroCall("$Inv", self ctxt at)))))) ])));
-            !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                                !!(MacroCall("$Inv", self ctxt at))))));
-          ])
+     !!([],
+        [ !!(FieldAssignS(
+            (self ctxt Source.no_region, !!id),
+            (!!(AddE(!!(FldAcc (self ctxt (s.at), !!id)),
+                     intLitE Source.no_region 1)))));
+          !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
```

Comment:
This is the relevant change, sorry for the other noise...

Reply:
`!!` still marks up the `await`, but `!@` points at the `async`.

Reply:
Right but doesn't the exhale belong to the await, not the async? This is when you need to establish the invariant before suspending. That was my thinking at least. But I don't know why it doesn't highlight the await rather than default to the unknow region.

Reply:
I think the `exhale` is due to the `async` being filed into the continuation table. But you are right, that only the `await` opens up the context switch so that a _havoc_ can occur.